packages/beacon-node/src/api/impl/beacon/state/utils.ts

```diff
@@ -1,53 +1,31 @@
 import {routes} from "@lodestar/api";
 import {FAR_FUTURE_EPOCH, GENESIS_SLOT} from "@lodestar/params";
 import {BeaconStateAllForks, PubkeyIndexMap} from "@lodestar/state-transition";
-import {BLSPubkey, Epoch, phase0, ValidatorIndex} from "@lodestar/types";
+import {BLSPubkey, Epoch, phase0, RootHex, Slot, ValidatorIndex} from "@lodestar/types";
 import {fromHex} from "@lodestar/utils";
-import {IBeaconChain, StateGetOpts} from "../../../../chain/index.js";
+import {IForkChoice} from "@lodestar/fork-choice";
+import {IBeaconChain} from "../../../../chain/index.js";
 import {ApiError, ValidationError} from "../../errors.js";
-import {isOptimisticBlock} from "../../../../util/forkChoice.js";
 
-export async function resolveStateId(
-  chain: IBeaconChain,
-  stateId: routes.beacon.StateId,
-  opts?: StateGetOpts
-): Promise<{state: BeaconStateAllForks; executionOptimistic: boolean; finalized: boolean}> {
-  const stateRes = await resolveStateIdOrNull(chain, stateId, opts);
-  if (!stateRes) {
-    throw new ApiError(404, `No state found for id '${stateId}'`);
-  }
-
-  return stateRes;
-}
-
-async function resolveStateIdOrNull(
-  chain: IBeaconChain,
-  stateId: routes.beacon.StateId,
-  opts?: StateGetOpts
-): Promise<{state: BeaconStateAllForks; executionOptimistic: boolean; finalized: boolean} | null> {
+export function resolveStateId(forkChoice: IForkChoice, stateId: routes.beacon.StateId): RootHex | Slot {
   if (stateId === "head") {
-    // TODO: This is not OK, head and headState must be fetched atomically
-    const head = chain.forkChoice.getHead();
-    const headState = chain.getHeadState();
-    return {state: headState, executionOptimistic: isOptimisticBlock(head), finalized: false};
```

Comment:
@wemeetagain wondering here, we got the head state from `getHeadState()` which called `getClosestHeadState`

so we would previouly do this

https://github.com/ChainSafe/lodestar/blob/fed08fe5107a809d8862494ae6a86ebc6f2f0ee9/packages/beacon-node/src/chain/regen/queued.ts#L142

which queried both `blockStateCache` and `checkpointStateCache`



while now we would just query `blockStateCache`

https://github.com/ChainSafe/lodestar/blob/fed08fe5107a809d8862494ae6a86ebc6f2f0ee9/packages/beacon-node/src/chain/regen/queued.ts#L79-L80

Reply:
@nflaig , the change actually made it better. Previously we also queried the head state pulled up to next epoch in the `checkpointStateCache` which is only necessary for block processing while the "getHeadState()" api only wants the real head state. So it's better to only query `blockStateCache` in this case